src/gc-alloc-profiler.cpp

```diff
@@ -0,0 +1,169 @@
+// This file is a part of Julia. License is MIT: https://julialang.org/license
+
+#include "gc-alloc-profiler.h"
+
+#include "julia_internal.h"
+#include "gc.h"
+
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+using std::unordered_map;
+using std::string;
+using std::vector;
+
+struct RawBacktrace {
+    jl_bt_element_t *data;
+    size_t size;
+};
+
+struct RawAlloc {
+    jl_datatype_t *type_address;
+    RawBacktrace backtrace;
+    size_t size;
+};
+
+struct PerThreadAllocProfile {
+    vector<RawAlloc> allocs;
+    unordered_map<size_t, size_t> type_address_by_value_address;
+    unordered_map<size_t, size_t> frees_by_type_address;
+
+    size_t alloc_counter;
+    size_t last_recorded_alloc;
+};
+
+struct AllocProfile {
+    int skip_every;
+
+    vector<PerThreadAllocProfile> per_thread_profiles;
+};
+
+struct CombinedResults {
+    vector<RawAlloc> combined_allocs;
+    vector<FreeInfo> combined_frees;
+};
+
+// == global variables manipulated by callbacks ==
+
+AllocProfile g_alloc_profile;
+int g_alloc_profile_enabled = false;
+CombinedResults g_combined_results; // will live forever
+
+// === stack stuff ===
+
+RawBacktrace get_raw_backtrace() {
+    static jl_bt_element_t static_bt_data[JL_MAX_BT_SIZE];
+
+    size_t bt_size = rec_backtrace(static_bt_data, JL_MAX_BT_SIZE, 2);
+
+    size_t bt_bytes = bt_size * sizeof(jl_bt_element_t);
+    jl_bt_element_t *bt_data = (jl_bt_element_t*) malloc(bt_bytes);
+    memcpy(bt_data, static_bt_data, bt_bytes);
+
+    return RawBacktrace{
+        bt_data,
+        bt_size
+    };
+}
+
+// == exported interface ==
+
+JL_DLLEXPORT void jl_start_alloc_profile(int skip_every) {
+    g_alloc_profile = AllocProfile{skip_every};
+
+    for (int i = 0; i < jl_n_threads; i++) {
+        g_alloc_profile.per_thread_profiles.push_back(PerThreadAllocProfile{});
+    }
+
+    g_alloc_profile_enabled = true;
+}
+
+extern "C" {  // Needed since the function doesn't take any arguments.
```

Comment:
I'm not totally sure, but I _think_ you should put this around your entire implementation of the exported interface (so lines 72 to the end).

I think it happens to work with how you have it right now, but i'm not sure if that's defined by the standards or undefined and happens to work.. I think the correct thing is that your _implementations_ of these functions should match how they were declared (which is with `extern "C"` linkage).

Reply:
resolved in https://github.com/vilterp/julia/pull/10